hack/generate/changelog/gen-changelog.go

```diff
@@ -0,0 +1,331 @@
+	}
+	if !haveMigrations {
+		bb.WriteString("There are no migrations for this release! :tada:\n\n")
+	}
+
+	if err := ioutil.WriteFile(c.File, bytes.TrimSuffix(bb.Bytes(), []byte("\n")), 0644); err != nil {
+		return fmt.Errorf("could not write migration guide: %v", err)
+	}
+	return nil
+}
+
+type fragment struct {
+	Entries []entry `yaml:"entries"`
+}
+
+type entry struct {
+	Description string     `yaml:"description"`
+	Kind        entryKind  `yaml:"kind"`
+	Breaking    bool       `yaml:"breaking"`
+	Migration   *migration `yaml:"migration,omitempty"`
+	PullRequest *uint      `yaml:"pull_request_override,omitempty"`
+}
+
+type entryKind string
+
+const (
+	addition    entryKind = "addition"
+	change      entryKind = "change"
+	removal     entryKind = "removal"
+	deprecation entryKind = "deprecation"
+	bugfix      entryKind = "bugfix"
+)
+
+func (k entryKind) toHeader() string {
+	switch k {
+	case addition:
+		return "### Additions"
+	case change:
+		return "### Changes"
+	case removal:
+		return "### Removals"
+	case deprecation:
+		return "### Deprecations"
+	case bugfix:
+		return "### Bug Fixes"
+	default:
+		panic("invalid entry kind; NOTE TO DEVELOPERS: check entryKind.validate")
+	}
+}
+
+type migration struct {
+	Header string `yaml:"header"`
+	Body   string `yaml:"body"`
+}
+
+type config struct {
+	File    string
+	Title   string
+	Entries []entry
+}
+
+func (f *fragment) validate() error {
+	for i, e := range f.Entries {
+		if err := e.validate(); err != nil {
+			return fmt.Errorf("entry[%d] invalid: %v", i, err)
+		}
+	}
+	return nil
+}
+
+func (e *entry) validate() error {
+	if err := e.Kind.validate(); err != nil {
+		return fmt.Errorf("invalid kind: %v", err)
+	}
+
+	if e.Breaking && e.Kind != change && e.Kind != removal {
+		return fmt.Errorf("breaking changes can only be kind %q or %q, got %q", change, removal, e.Kind)
+	}
+
+	if e.Breaking && e.Migration == nil {
+		return fmt.Errorf("breaking changes require migration descriptions")
+	}
+
+	if e.Migration != nil {
+		if err := e.Migration.validate(); err != nil {
+			return fmt.Errorf("invalid migration: %v", err)
+		}
+	}
+	return nil
+}
+
+func (e entry) toChangelogString() string {
+	text := strings.Trim(e.Description, "\n")
+	if e.Breaking {
+		text = fmt.Sprintf("**Breaking change**: %s", text)
+	}
+	if !strings.HasSuffix(text, ".") && !strings.HasSuffix(text, "!") {
+		text = fmt.Sprintf("%s.", text)
+	}
+	if e.PullRequest != nil {
+		text = fmt.Sprintf("%s (%s)", text, e.pullRequestLink())
+	}
+	return fmt.Sprintf("- %s", text)
+}
+
+func (e entry) pullRequestLink() string {
+	const repo = "github.com/operator-framework/operator-sdk"
+	return fmt.Sprintf("[#%d](https://%s/pull/%d)", *e.PullRequest, repo, *e.PullRequest)
+}
+
+func getPullRequest(filename string) (uint, error) {
+	args := fmt.Sprintf("log --follow --pretty=format:%%s --diff-filter=A --find-renames=40%% %s", filename)
+	line, err := exec.Command("git", strings.Split(args, " ")...).CombinedOutput()
```

Comment:
You could probably break this part out to unit test this function, although I don't expect PR commit message structure to change.

Reply:
You're saying we could have one function to retrieve the original commit message and another function that parses the PR number from it?

Reply:
Updated to split into separate functions. No unit tests (yet)

Reply:
Refactored and added a bunch of unit tests. Package now has 86.5% coverage :slightly_smiling_face: